explorer/interpreter/matching_impl_set.h

```diff
@@ -0,0 +1,121 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_EXPLORER_INTERPRETER_MATCHING_IMPL_SET_H_
+#define CARBON_EXPLORER_INTERPRETER_MATCHING_IMPL_SET_H_
+
+#include <vector>
+
+#include "common/ostream.h"
+#include "explorer/ast/declaration.h"
+#include "explorer/common/nonnull.h"
+#include "explorer/interpreter/impl_scope.h"
+#include "explorer/interpreter/value.h"
+#include "llvm/ADT/DenseMap.h"
+
+namespace Carbon {
+
+// A set of impl matches that we're currently performing. Each `Match`
+// represents an attempt to match a `type as interface` query against an `impl`
+// declaration. This is used to detect and reject non-termination when impl
+// matching recursively triggers further impl matching.
+//
+// The approach we use is to count the number of times each "label" appears
+// within the type and interface, where a label is the name of a declared
```

Comment:
```suggestion
// The approach we use is to count the number of times each "Label" appears
// within the type and interface, where a Label is the name of a declared
```

Use the capital Label to make it clear it's the type?

Reply:
My intent for this paragraph was to describe the language rule being implemented, not the implementation details of the class. I didn't actually mean the type `Label` here. I've rephrased to try to make this clearer.